src/main/java/org/tugraz/sysds/runtime/io/FrameReaderJSONLParallel.java

```diff
@@ -0,0 +1,128 @@
+package org.tugraz.sysds.runtime.io;
+
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.io.LongWritable;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.mapred.*;
+import org.tugraz.sysds.common.Types;
+import org.tugraz.sysds.hops.OptimizerUtils;
+import org.tugraz.sysds.runtime.matrix.data.FrameBlock;
+import org.tugraz.sysds.runtime.util.CommonThreadPool;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+
+public class FrameReaderJSONLParallel extends FrameReaderJSONL {
+
+
+    @Override
+    protected void readJSONLFrameFromHDFS(Path path, JobConf jobConf, FileSystem fileSystem, FrameBlock dest,
+                                          Types.ValueType[] schema, Map<String, Integer> schemaMap)
+            throws IOException{
+
+        int numThreads = OptimizerUtils.getParallelTextReadParallelism();
+
+        TextInputFormat inputFormat = new TextInputFormat();
+        inputFormat.configure(jobConf);
+        InputSplit[] splits = inputFormat.getSplits(jobConf, numThreads);
+        splits = IOUtilFunctions.sortInputSplits(splits);
+
+        try{
+            ExecutorService executorPool = CommonThreadPool.get(Math.min(numThreads, splits.length));
+
+            //compute num rows per split
+            ArrayList<CountRowsTask> countRowsTasks = new ArrayList<>();
+            for (InputSplit split : splits){
+                countRowsTasks.add(new CountRowsTask(split, inputFormat, jobConf));
+            }
+            List<Future<Long>> futureCountList = executorPool.invokeAll(countRowsTasks);
+
+            //compute row offset per split via cumsum on row counts
+            long offset = 0;
+            List<Long> offsets = new ArrayList<>();
+            for( Future<Long> count : futureCountList ) {
+                offsets.add(offset);
+                offset += count.get();
+            }
+
+            //read individual splits
+            ArrayList<ReadRowsTask> readRowsTasks = new ArrayList<>();
+            for( int i=0; i<splits.length; i++ ) {
+                readRowsTasks.add(new ReadRowsTask(splits[i], inputFormat, jobConf, dest, schemaMap, offsets.get(i).intValue()));
+            }
+            List<Future<Object>> rret = executorPool.invokeAll(readRowsTasks);
+            executorPool.shutdown();
+
+            //error handling
+            for( Future<Object> read : rret )
+                read.get();
+
+        }catch (Exception e) {
+            throw new IOException("Failed parallel read of JSONL input.", e);
+        }
+
+    }
+
+    private static class CountRowsTask implements Callable<Long> {
```

Comment:
If this implementation is equivalent to the CSV `CountRowsTasks`, maybe we can put this into `IOUtilFunctions`.